README.md
# Command line Zello player

This is a command line tool to play audio files on a Zello channel.

It uses [TS Zello](https://github.com/OnkelTem/ts-zello) library.

The tool is currently under development and is not ready for "production".
There are things to do first, see the [TODO](#todo) section below.

## Installation

### Linux (Ubuntu)

Install the player:

```
$ npm i -g ts-zello-player
```

Install [FFmpeg](https://ffmpeg.org/):

```
$ sudo apt-get install ffmpeg
```

Install `build-essential` package:

```
$ sudo apt-get install build-essential
```

Install `Roboto` font:

```
$ sudo apt-get install fonts-roboto
```

### Other OSes

Haven't tested or tried. Contributions are welcome.

## Configuration

Create a json file `credentials.json` in the following format: 

```json
{
    "username": "<User Name>",
    "password": "<Password>",
    "auth_token": "<Auth Token>",
    "channel": "<Channel Name>"
}
```

See [Zello authentication docs](https://github.com/zelloptt/zello-channel-api/blob/master/AUTH.md) 
on how to get the token for your account.

## Quick start

Play a file on the default channel from `credentials.json`:

```
$ ts-zello-player "file.mp3" 
```

Play a YouTube video on another channel:

```
$ ts-zello-player -c ChannelName "https://www.youtube.com/watch?v=0hLnMDJ-gV4"
```

Play a file 
- from 35th second 
- at 0.75 speed
- at 0.5 volume 
- with bitrate of 32 Kbps
- on the channel "ChannelName":

```
$ ts-zello-player -b 32 -t 0.75 -l 0.5 -c ChannelName file.mp3 35
```

For the full list of options see below.

## Usage

```
$ ts-zello-player --help

Usage: ts-zello-player [options] <file> [start]
Usage: ts-zello-player [options] <youtube link> [start]

Positionals:
  target  A target to play, either file path or a YouTube link          [string]
  start   Start playback at                                             [string]

Options:
      --version      Show version number                               [boolean]
  -b, --bitrate      Bitrate in Kbps                      [number] [default: 48]
  -l, --volume       Volume factor                                      [number]
  -r, --rate         (re)Sampling rate
                    [choices: 8000, 12000, 16000, 24000, 48000] [default: 48000]
  -t, --tempo        Tempo factor                                       [number]
  -f, --frame        Frame size  [choices: 2.5, 5, 10, 20, 40, 60] [default: 20]
      --credentials  Credentials file path[string] [default: "credentials.json"]
  -c, --channel      Zello channel to connect to                        [string]
  -p, --preview      Show recording image preview     [boolean] [default: false]
  -i, --info         Show recording text information  [boolean] [default: false]
      --normalize    Normalize sound with FFmpeg "loudnorm" filter
                                                      [boolean] [default: false]
      --compress     Compress sound with FFmpeg "acompressor" filter
                                                      [boolean] [default: false]
  -h, --help         Show help                                         [boolean]

Examples:
  ts-zello-player file.mp3                  play file.mp3
  ts-zello-player https://www.youtube.com/  play audio from the YouTube video
  watch?v=0hLnMDJ-gV4

```

## TODO

- Finish this README
- Record a short helping screencast 
- Add _retry strategy_ selection to the command line options


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.6](https://github.com/OnkelTem/ts-zello-player/compare/v0.0.5...v0.0.6) (2021-04-23)


### Bug Fixes

* **dev:** added proper paths to assets ([3a22b6a](https://github.com/OnkelTem/ts-zello-player/commit/3a22b6a761390fb5ee64a68dd5c362d67170c142))

### [0.0.5](https://github.com/OnkelTem/ts-zello-player/compare/v0.0.4...v0.0.5) (2021-04-23)


### Bug Fixes

* **dev:** moved SVG imagery into 'src' directory, updated README ([441a536](https://github.com/OnkelTem/ts-zello-player/commit/441a536dc7293c7b7dccfe8f304eeadc51f9aa85))

### [0.0.4](https://github.com/OnkelTem/ts-zello-player/compare/v0.0.3...v0.0.4) (2021-04-23)


### Bug Fixes

* **readme:** updated USAGE ([e7af264](https://github.com/OnkelTem/ts-zello-player/commit/e7af2647fae6365b767179f38aee33baa5c16689))

### [0.0.3](https://github.com/OnkelTem/ts-zello-player/compare/v0.0.2...v0.0.3) (2021-04-23)


### Features

* **features:** added channel selection, (radio) streams playback, infos and previews etc ([2711d59](https://github.com/OnkelTem/ts-zello-player/commit/2711d590575bc1f95f1c937d57d05968ae0ae64c))

### [0.0.2](https://github.com/OnkelTem/ts-zello-player/compare/v0.0.1...v0.0.2) (2021-04-18)


### Features

* **cli:** added cli support ([dcee12c](https://github.com/OnkelTem/ts-zello-player/commit/dcee12ca78ff6c25136685754677a526fe356887))

### 0.0.1 (2021-04-10)


src/cli.ts
import { samplingRates, frameSizes, CommandLogonRequest, LoggerLevel, loggerLevels } from 'ts-zello';
import yargs from 'yargs';
import findUp from 'find-up';
import { hideBin } from 'yargs/helpers';
import { existsSync, readFileSync } from 'fs';

const configPath = findUp.sync(['.ts-zello-player', '.ts-zello-player.json']);
const config = configPath != null ? JSON.parse(readFileSync(configPath, 'utf8')) : {};

const options = {
  bitrate: {
    alias: 'b',
    default: 32,
    description: 'Bitrate in Kbps',
    number: true,
    min: 4,
    max: 96,
  },
  volume: {
    alias: 'l',
    default: 0.3,
    description: 'Volume factor',
    number: true,
    min: 0.01,
    max: 2,
  },
  rate: {
    alias: 'r',
    default: 48000,
    description: '(re)Sampling rate',
    choices: samplingRates,
  },
  tempo: {
    alias: 't',
    default: 1,
    description: 'Tempo factor',
    min: 0.5,
    max: 2,
  },
  frame: {
    alias: 'f',
    default: 20,
    description: 'Frame size',
    choices: frameSizes,
  },
  credentials: {
    default: 'credentials.json',
    description: 'Credentials file path',
    normalize: true,
  },
  channel: {
    alias: 'c',
    description: 'Zello channel to connect to',
    type: 'string',
  },
} as const;
type OptionName = keyof typeof options;

function checkOptionRange(optionName: OptionName, value: number) {
  const option = options[optionName];
  if ('min' in option && 'max' in option) {
    if (value != null && (value > option.max || value < option.min)) {
      throw new Error(`${option.description} (${optionName}) must be between ${option.min} and ${option.max}`);
    }
  }
}

const argv = yargs(hideBin(process.argv))
  .options(options)
  .config(config)
  .demandCommand(1)
  .command('$0 <target> [start]', 'the default command', (yargs) =>
    yargs
      .positional('target', {
        describe: 'A target to play, either file path or a YouTube link',
        type: 'string',
      })
      .positional('start', {
        description: 'Start playback at',
        type: 'string',
      })
      .check((argv) => {
        checkOptionRange('bitrate', argv.bitrate);
        checkOptionRange('volume', argv.volume);
        checkOptionRange('tempo', argv.tempo);
        return true;
      })
      .fail(function (msg, err, yargs) {
        console.log(yargs.help());
        console.error('\n\x1b[31m%s\x1b[0m', msg);
        process.exit(1);
      })
      .usage('Usage: $0 [options] <file> [start]')
      .usage('Usage: $0 [options] <youtube link> [start]')
      .example('$0 file.mp3', 'play file.mp3')
      .example('$0 https://www.youtube.com/watch?v=0hLnMDJ-gV4', 'play audio from the YouTube video'),
  )
  .count('verbose')
  .alias('v', 'verbose')
  .default('verbose', 0)
  .coerce('credentials', function (arg): CommandLogonRequest {
    if (!existsSync(arg)) {
      throw new Error(`Credentials file not found: ${arg}`);
    }
    const raw = readFileSync(arg, 'utf8');
    return JSON.parse(raw);
  })
  .coerce('verbose', function (arg) {
    return getLoggerLevel(arg);
  })
  .help('h')
  .alias('h', 'help').argv;

export function getLoggerLevel(count: number): LoggerLevel {
  // Zero amount of `-v` should result to 'info' log level, which corresponds to index = 3
  const index = count + 3;
  if (index > loggerLevels.length - 1) {
    return loggerLevels[loggerLevels.length - 1];
  } else {
    return loggerLevels[index];
  }
}

export default argv;


src/main.ts
import fs from 'fs';
import ytdl from 'ytdl-core';
import pino from 'pino';
import { Readable } from 'stream';
import { basename } from 'path';
import NodeID3 from 'node-id3';
import pEvent from 'p-event';
import fetch from 'node-fetch';
import sharp from 'sharp';
import zello, {
  Zello,
  DEFAULT_ZELLO_OPTIONS,
  getAutoDecodeStream,
  FileStreamOptions,
  SamplingRate,
  FrameSize,
  CommandLogonRequest,
  LoggerLevel,
  DataWaitPassThroughStream,
} from 'ts-zello';

const ZELLO_SERVER = 'wss://zello.io/ws';

type MainParams = {
  bitrateKbps: number;
  volumeFactor?: number;
  samplingRate: SamplingRate;
  tempoFactor?: number;
  frameSize: FrameSize;
  credentials: CommandLogonRequest;
  channel?: string;
  target: string;
  startAt?: string;
  logLevel: LoggerLevel;
  normalizer?: boolean;
  compressor?: boolean;
  details: {
    text: boolean;
    image: boolean;
  };
};

export async function main({
  bitrateKbps,
  volumeFactor,
  samplingRate,
  tempoFactor,
  frameSize,
  credentials,
  channel,
  target,
  startAt,
  logLevel,
  normalizer,
  compressor,
  details,
}: MainParams) {
  const pinoLogger = pino({ ...DEFAULT_ZELLO_OPTIONS.logger, level: logLevel });

  // @see: https://stackoverflow.com/a/3726073/1223483
  const youTubeLinkRegEx = /^http(?:s?):\/\/(?:www\.)?(?:music\.)?youtu(?:be\.com\/watch\?v=|\.be\/)([\w\-\_]*)(&(amp;)?‌​[\w\?‌​=]*)?/;

  const urlRegEx = /^http(?:s?):\/\//;

  let stream: Readable | undefined = undefined;
  let z: Zello;

  const cred = credentials!;

  // Override channel from the cli
  if (channel != null) {
    cred.channel = channel;
  }

  const fileStreamOptions: FileStreamOptions = {
    samplingRate,
    volumeFactor,
    tempoFactor,
    normalizer,
    compressor,
  };

  async function start() {
    let image: Buffer;
    let detailsText: string;
    try {
      if (target.match(youTubeLinkRegEx)) {
        // YouTube Link
        pinoLogger.info('YouTube link detected');
        pinoLogger.debug('Requesting YouTube video metadata');

        const videoInfo = await ytdl.getBasicInfo(target);
        if (details.text) {
          detailsText = makeInfoLineFromYouTubeInfo(videoInfo);
        }
        if (details.image) {
          image = await makeYouTubePreview(videoInfo);
        }
        pinoLogger.debug('Requesting YouTube stream');
        // Check to see the length of the video and is offset available.
        // Wait! Maybe it's already there?
        const ytdlOptions: ytdl.downloadOptions = {
          quality: 'highestaudio',
          ...(startAt != null ? { begin: startAt } : null),
        };
        const audio = ytdl(target, ytdlOptions);
        stream = audio.pipe(getAutoDecodeStream(pinoLogger, fileStreamOptions));
      } else if (target.match(urlRegEx)) {
        // Just a link
        pinoLogger.info('URL detected');
        pinoLogger.debug('Requesting URL...');
        const audio = (await fetch(target)).body;
        stream = audio.pipe(getAutoDecodeStream(pinoLogger, fileStreamOptions));
      } else {
        // File
        pinoLogger.info('Local file detected');
        if (!fs.existsSync(target)) {
          throw new Error(`File not found: "${target}"`);
        }
        const tags = NodeID3.read(target);
        if (details.text) {
          detailsText = makeInfoLineFromTags(tags) || '🎵\u00A0' + basename(target);
        }
        if (details.image) {
          if (tags.image != null && typeof tags.image === 'object') {
            image = tags.image.imageBuffer;
          }
        }
        stream = fs.createReadStream(target).pipe(getAutoDecodeStream(pinoLogger, { ...fileStreamOptions, startAt }));
      }
    } catch (err) {
      pinoLogger.error(err, 'ERROR');
      process.exit(2);
    }

    if (stream == null) {
      pinoLogger.error("Stream couldn't initialize");
      process.exit(3);
    }

    // This part is not required (it's not inside sendAudio macro),
    // but it decreases the delay between now-playing info and the audio transmission.
    pinoLogger.info('Waiting for data...');
    stream = stream.pipe(new DataWaitPassThroughStream());
    await pEvent(stream, 'dataIsReady');
    pinoLogger.debug('Data is ready.');

    z = await zello(ZELLO_SERVER, { logger: pinoLogger, name: 'ts-zello-player' });
    try {
      await z.ctl.run(function* ({ macros, commands }) {
        yield macros.login(cred);
        if (image != null) {
          yield macros.sendImage(image);
        }
        if (detailsText != null) {
          yield commands.sendTextMessage({ text: detailsText });
        }
        yield macros.sendAudio(stream!, {
          transcode: { samplingRate, frameSize, bitrateKbps, channels: 1 },
        });
      });
    } catch (err) {
      pinoLogger.error(err, 'ERROR');
    }
    pinoLogger.info('==> Closing socket');
    await z.ctl.close();
    pinoLogger.info('==> Socket is closed');
  }

  async function shutdown() {
    if (z && z.ctl.status() === 'OPEN') {
      pinoLogger.warn('Closing...');
      await stream!.destroy();
      await z.ctl.close();
    }
  }

  process.on('SIGINT', async function () {
    pinoLogger.warn('Received SIGINT: Stopped by user');
    await shutdown();
    process.exit();
  });

  await start();
  pinoLogger.warn('==> end of main function');
}

function makeInfoLineFromTags({ artist, album, year, trackNumber, title }: NodeID3.Tags): string {
  const _ = '\u00A0';
  return [
    artist ? `${artist}` : null,
    album ? `💿${_}${album} ${year ? `${_}(${year})` : ''}` : null,
    trackNumber || title ? `🎵${_}${trackNumber ? `${trackNumber} ` : ''}${title ? `- ${title}` : ''}` : null,
  ]
    .filter((item) => item != null)
    .join('\n');
}

function secondsToTimeFmt(seconds: number) {
  const secReminder = seconds % 60;
  const minutes = (seconds - secReminder) / 60;
  const minutesReminder = minutes % 60;
  const hours = (minutes - minutesReminder) / 60;
  let output: string = '';
  if (hours > 0) {
    output = `${hours.toString().padStart(2, '0')}:`;
  }
  output += `${minutes.toString().padStart(2, '0')}:${secReminder.toString().padStart(2, '0')}`;
  return output;
}

function numberFormat(n: number) {
  if (n > 1000000000) {
    return `${Math.round(n / 1000000000)}B`;
  }
  if (n > 1000000) {
    return `${Math.round(n / 1000000)}M`;
  }
  if (n > 1000) {
    return `${Math.round(n / 1000)}k`;
  }
  return n.toLocaleString();
}

function makeInfoLineFromYouTubeInfo({
  videoDetails: { title, dislikes, likes, ownerChannelName, uploadDate, viewCount, lengthSeconds },
}: ytdl.videoInfo): string {
  const _ = '\u00A0';
  const lengthFmt = secondsToTimeFmt(parseInt(lengthSeconds));
  let viewsFmt: string = viewCount;
  if (viewCount.match(/^\d+$/)) {
    viewsFmt = numberFormat(parseInt(viewCount));
  }
  const likesFmt = numberFormat(likes ?? 0);
  const dislikesFmt = numberFormat(dislikes ?? 0);
  return [
    title,
    `👁${_}${viewsFmt}${_}${_}👍${_}${likesFmt}${_}${_}👎${_}${dislikesFmt}${_}${_}🕒${_}${lengthFmt}`,
    `ⓘ Channel: ${_}${ownerChannelName}${_}${_}⬆${_}${uploadDate}`,
  ]
    .filter((item) => item != null)
    .join('\n');
}

function findThumbnail(thumbnails: ytdl.thumbnail[]): string {
  let maxWidth = 0;
  let maxIndex = 0;
  for (let i = 0; i < thumbnails.length; i++) {
    if (thumbnails[i].width > maxWidth) {
      maxWidth = thumbnails[i].width;
      maxIndex = i;
    }
  }
  return thumbnails[maxIndex].url;
}

async function makeYouTubePreview(videoInfo: ytdl.videoInfo): Promise<Buffer> {
  const {
    videoDetails: { title, dislikes, likes, ownerChannelName, uploadDate, viewCount, lengthSeconds },
  } = videoInfo;
  const lengthFmt = secondsToTimeFmt(parseInt(lengthSeconds));
  let viewsFmt: string = viewCount;
  if (viewCount.match(/^\d+$/)) {
    viewsFmt = numberFormat(parseInt(viewCount));
  }
  const likesFmt = numberFormat(likes ?? 0);
  const dislikesFmt = numberFormat(dislikes ?? 0);

  const thumbnailUrl = findThumbnail(videoInfo.videoDetails.thumbnails);
  let imageData = await (await fetch(thumbnailUrl)).buffer();
  imageData = await sharp(imageData).resize({ width: 800, height: 450 }).toBuffer();

  let svgPageData = fs.readFileSync(__dirname + '/../src/assets/drawing-opt.svg', 'utf8');
  svgPageData = svgPageData.replace('#videoTitle', title);
  svgPageData = svgPageData.replace('#channelName', ownerChannelName);
  svgPageData = svgPageData.replace('#viewsCount', viewsFmt);
  svgPageData = svgPageData.replace('#likes', likesFmt);
  svgPageData = svgPageData.replace('#dislikes', dislikesFmt);
  svgPageData = svgPageData.replace('#uploaded', uploadDate);
  const svgPageBuffer = Buffer.from(svgPageData);

  let svgTimeLabelData = fs.readFileSync(__dirname + '/../src/assets/time-label-opt.svg', 'utf8');
  svgTimeLabelData = svgTimeLabelData.replace('#time', lengthFmt);
  const svgTimeLabelBuffer = Buffer.from(svgTimeLabelData);

  return await sharp(svgPageBuffer)
    .composite([
      {
        input: imageData,
        left: 110,
        top: 0,
      },
      {
        input: svgTimeLabelBuffer,
        left: 850,
        top: 380,
      },
    ])
    .jpeg()
    .toBuffer();
}


src/index.ts
import { samplingRates, frameSizes, CommandLogonRequest, LoggerLevel, loggerLevels } from 'ts-zello';
import yargs from 'yargs';
import findUp from 'find-up';
import { hideBin } from 'yargs/helpers';
import { existsSync, readFileSync } from 'fs';
import { main } from './main';

const configPath = findUp.sync(['.ts-zello-player', '.ts-zello-player.json']);
const config = configPath != null ? JSON.parse(readFileSync(configPath, 'utf8')) : {};

const options = {
  bitrate: {
    alias: 'b',
    default: 48,
    description: 'Bitrate in Kbps',
    number: true,
    min: 4,
    max: 96,
  },
  volume: {
    alias: 'l',
    description: 'Volume factor',
    number: true,
    min: 0.01,
    max: 2,
  },
  rate: {
    alias: 'r',
    default: 48000,
    description: '(re)Sampling rate',
    choices: samplingRates,
  },
  tempo: {
    alias: 't',
    description: 'Tempo factor',
    number: true,
    min: 0.5,
    max: 2,
  },
  frame: {
    alias: 'f',
    default: 20,
    description: 'Frame size',
    choices: frameSizes,
  },
  credentials: {
    default: 'credentials.json',
    description: 'Credentials file path',
    normalize: true,
  },
  channel: {
    alias: 'c',
    description: 'Zello channel to connect to',
    type: 'string',
  },
  preview: {
    alias: 'p',
    default: false,
    description: 'Show recording image preview',
    type: 'boolean',
  },
  info: {
    alias: 'i',
    default: false,
    description: 'Show recording text information',
    type: 'boolean',
  },
  normalize: {
    default: false,
    description: 'Normalize sound with FFmpeg "loudnorm" filter',
    type: 'boolean',
  },
  compress: {
    default: false,
    description: 'Compress sound with FFmpeg "acompressor" filter',
    type: 'boolean',
  },
} as const;
type OptionName = keyof typeof options;

function checkOptionRange(optionName: OptionName, value: number) {
  const option = options[optionName];
  if ('min' in option && 'max' in option) {
    if (value != null && (value > option.max || value < option.min)) {
      throw new Error(`${option.description} (${optionName}) must be between ${option.min} and ${option.max}`);
    }
  }
}

yargs(hideBin(process.argv))
  .options(options)
  .config(config)
  .demandCommand(1)
  .command(
    '$0 <target> [start]',
    'the default command',
    (yargs) =>
      yargs
        .positional('target', {
          describe: 'A target to play, either file path or a YouTube link',
          type: 'string',
        })
        .positional('start', {
          description: 'Start playback at',
          type: 'string',
        })
        .check((argv) => {
          checkOptionRange('bitrate', argv.bitrate);
          if (argv.volume != null) {
            checkOptionRange('volume', argv.volume);
          }
          if (argv.tempo != null) {
            checkOptionRange('tempo', argv.tempo);
          }
          return true;
        })
        .fail(function (msg, err, yargs) {
          console.log(yargs.help());
          console.error('\n\x1b[31m%s\x1b[0m', msg);
          process.exit(1);
        })
        .usage('Usage: $0 [options] <file> [start]')
        .usage('Usage: $0 [options] <youtube link> [start]')
        .example('$0 file.mp3', 'play file.mp3')
        .example('$0 https://www.youtube.com/watch?v=0hLnMDJ-gV4', 'play audio from the YouTube video')
        .count('verbose')
        .alias('v', 'verbose')
        .default('verbose', 0)
        .coerce('credentials', function (arg): CommandLogonRequest {
          if (!existsSync(arg)) {
            throw new Error(`Credentials file not found: ${arg}`);
          }
          const raw = readFileSync(arg, 'utf8');
          return JSON.parse(raw);
        })
        .coerce('verbose', function (arg) {
          return getLoggerLevel(arg);
        }),
    async function (argv) {
      if (argv.credentials != null && argv.target != null) {
        await main({
          bitrateKbps: argv.bitrate,
          volumeFactor: argv.volume,
          samplingRate: argv.rate,
          tempoFactor: argv.tempo,
          frameSize: argv.frame,
          credentials: argv.credentials,
          channel: argv.channel,
          target: argv.target,
          startAt: argv.start,
          logLevel: argv.verbose!,
          normalizer: argv.normalize,
          compressor: argv.compress,
          details: {
            text: argv.info,
            image: argv.preview,
          },
        });
        process.exit(0);
      }
    },
  )
  .help('h')
  .alias('h', 'help').argv;

function getLoggerLevel(count: number): LoggerLevel {
  // Zero amount of `-v` should result to 'info' log level, which corresponds to index = 3
  const index = count + 3;
  if (index > loggerLevels.length - 1) {
    return loggerLevels[loggerLevels.length - 1];
  } else {
    return loggerLevels[index];
  }
}
